Environment/CTP_generator.py

```diff
@@ -1,103 +1,153 @@
-import jax 
+import jax
 import jax.numpy as jnp
-import jraph 
+import jraph
 import numpy as np
 from scipy.spatial import Delaunay
 from functools import partial
 import networkx as nx
 import matplotlib.pyplot as plt
 import os
 
-#Generate a random graph 
-#With the same number of nodes, the total number of edges can be different.
-#Cannot currently jit this function because of the conversion to numpy array for Delauney
-def generate_graph(n_nodes: int, key: int, use_edge_weights=True, grid_size=10,prop_stoch=0.4
-                   ) -> tuple[jraph.GraphsTuple, int, int]:
+
+# Generate a random graph
+# With the same number of nodes, the total number of edges can be different.
+# Cannot currently jit this function because of the conversion to numpy array for Delauney
+def generate_graph(
+    n_nodes: int, key: int, use_edge_weights=True, grid_size=10, prop_stoch=0.4
+) -> tuple[jraph.GraphsTuple, int, int]:
 
     xmax = grid_size
     ymax = grid_size
-    #Generate random points in the grid
+    # Generate random points in the grid
     node_pos = jax.random.choice(key, xmax * ymax, (n_nodes,), replace=False)
-    grid_nodes = jax.vmap(_convert_to_grid,in_axes=(0,None))(node_pos,ymax) 
+    grid_nodes = jax.vmap(_convert_to_grid, in_axes=(0, None))(node_pos, ymax)
     grid_nodes_jax = jnp.array(grid_nodes).T
 
-    #Choose orign and goal nodes
-    origin,goal = find_goal_and_origin(grid_nodes_jax)
+    # Choose orign and goal nodes
+    origin, goal = find_goal_and_origin(grid_nodes_jax)
 
     # Apply Delauney triangulation to get edges
     delaunay = Delaunay(grid_nodes_jax)
     simplices = delaunay.simplices
     simplices_jax = jnp.array(simplices)
-    
+
     # Extract edges from the simplices, treat them as undirected edges
-    # Remove duplicate edges. Sender node number always > receiver node number now. 
+    # Remove duplicate edges. Sender node number always > receiver node number now.
```

Comment:
We might want to separate goal/origin selection and graph generation entirely, since in the multi-agent case the agents will have different goals but the same underlying graph